File: backend/app/api/utils/image_url_utils.py
"""
Utilities for normalizing image URLs to support deduplication across size variants.
Common CDN patterns: Shopify (width=, height=), imgix (w=, h=), Cloudinary, etc.
"""

from urllib.parse import parse_qs, urlencode, urlparse, urlunparse

# Query params that specify image dimensions (strip for canonical dedup)
SIZE_PARAMS = frozenset({"width", "height", "w", "h", "size", "resize", "dpr", "q", "quality", "format", "fit", "crop"})

# Prefer this width when fetching high-res (Shopify supports up to 5760)
PREFERRED_WIDTH = 5760


def get_canonical_image_url(url: str) -> str:
    """
    Normalize an image URL to a canonical form for deduplication.
    Strips size-related query params so that ...&width=416 and ...&width=3000
    map to the same canonical URL.
    """
    if not url or not url.strip():
        return ""
    try:
        parsed = urlparse(url.strip())
        if not parsed.scheme or not parsed.netloc:
            return url
        # Normalize scheme to https
        scheme = "https" if parsed.scheme in ("http", "https") else parsed.scheme
        netloc = parsed.netloc.lower().lstrip("www.")
        # Parse query and remove size params (keep v=, etc. for identity)
        params = parse_qs(parsed.query, keep_blank_values=False)
        filtered = {k: v for k, v in params.items() if k.lower() not in SIZE_PARAMS}
        new_query = urlencode(filtered, doseq=True)
        # Rebuild URL
        return urlunparse((scheme, netloc, parsed.path or "/", parsed.params, new_query, ""))
    except Exception:
        return url


def get_high_res_image_url(url: str) -> str:
    """
    Transform an image URL to request high resolution.
    Adds or replaces width param with a high value (e.g. Shopify supports 5760).
    Use this when fetching to store the best quality.
    """
    if not url or not url.strip():
        return url
    try:
        parsed = urlparse(url.strip())
        if not parsed.scheme or not parsed.netloc:
            return url
        scheme = "https" if parsed.scheme in ("http", "https") else parsed.scheme
        netloc = parsed.netloc.lower()
        params = parse_qs(parsed.query, keep_blank_values=True)
        # Remove existing size params
        params = {k: v for k, v in params.items() if k.lower() not in SIZE_PARAMS}
        # Add width for high-res (Shopify, many CDNs use width=)
        params["width"] = [str(PREFERRED_WIDTH)]
        new_query = urlencode(params, doseq=True)
        return urlunparse((scheme, netloc, parsed.path or "/", parsed.params, new_query, ""))
    except Exception:
        return url

File: chrome-extension/src/utils/imageUrlUtils.ts
/**
 * Normalize image URLs across CDN size variants (Shopify, imgix and similar) so
 * the same image dedupes to one entry.
 */

const SIZE_PARAMS = new Set([
  "width",
  "height",
  "w",
  "h",
  "size",
  "resize",
  "dpr",
  "q",
  "quality",
  "format",
  "fit",
  "crop",
]);

const PREFERRED_WIDTH = 5760;

/** Strip size params so every variant of an image shares one canonical URL. */
export function getCanonicalImageUrl(url: string): string {
  if (!url?.trim()) return "";
  try {
    const u = new URL(url.trim());
    const params = new URLSearchParams(u.search);
    for (const key of Array.from(params.keys())) {
      if (SIZE_PARAMS.has(key.toLowerCase())) {
        params.delete(key);
      }
    }
    const newSearch = params.toString();
    u.search = newSearch;
    u.protocol = "https:";
    u.hostname = u.hostname.toLowerCase().replace(/^www\./, "");
    return u.toString();
  } catch {
    return url;
  }
}

/** Request the highest resolution variant, for fetching an image to store. */
export function getHighResImageUrl(url: string): string {
  if (!url?.trim()) return url;
  try {
    const u = new URL(url.trim());
    const params = new URLSearchParams(u.search);
    for (const key of Array.from(params.keys())) {
      if (SIZE_PARAMS.has(key.toLowerCase())) {
        params.delete(key);
      }
    }
    params.set("width", String(PREFERRED_WIDTH));
    u.search = params.toString();
    u.protocol = "https:";
    return u.toString();
  } catch {
    return url;
  }
}
